File: Daily_Challenge/20230406_LC_1020.py
class Solution:
    def numEnclaves(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        # mark all 1s to 0s from the boundary
        def mark_boundary(grid, row, col):
            if row < 0 or row == m or col < 0 or col == n or grid[row][col] == 0:
                return
            grid[row][col] = 0
            mark_boundary(grid, row+1, col)
            mark_boundary(grid, row-1, col)
            mark_boundary(grid, row, col+1)
            mark_boundary(grid, row, col-1)
        
        
        for row in range(m):
            if grid[row][0] == 1:
                mark_boundary(grid, row, 0)
            if grid[row][n-1] == 1:
                mark_boundary(grid, row, n-1)
        
        for col in range(n):
            if grid[0][col] == 1:
                mark_boundary(grid, 0, col)
            if grid[m-1][col] == 1:
                mark_boundary(grid, m-1, col)
        ans = 0        
        # find the remaining 1s in the grid
        for row in range(m):
            ans += sum(grid[row])
        return ans
        """
        def dfs(grid, row, col):
            if row < 0 or row == m or col < 0 or col == n or grid[row][col] == 0:
                return 0
            grid[row][col] = 0
            t = dfs(grid, row-1, col)
            d = dfs(grid, row+1, col)
            r = dfs(grid, row, col+1)
            l = dfs(grid, row, col-1)
            return 1 + t + d + r + l
        
        for row in range(m):
            for col in range(n):
                if grid[row][col] == 1:
                    ans += dfs(grid, row, col)
        return ans
        
        """
        

File: Daily_Challenge/20230406_LC_1020.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& grid, int row, int col, int m, int n){
        if (row < 0 || row == m || col < 0 || col == n){
            return;
        }
        if (grid[row][col] != 1){
            return;
        }
        grid[row][col] = -1;
        dfs(grid, row+1, col, m, n);
        dfs(grid, row-1, col, m, n);
        dfs(grid, row, col+1, m, n);
        dfs(grid, row, col-1, m, n);
        return;
    }
    int numEnclaves(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        int ans = 0;
        // mark the 1 that can be reached from the boundary
        for (int row = 0; row < m; row++){
            if (grid[row][0] == 1){
                dfs(grid, row, 0, m, n);
            }
            if (grid[row][n-1] == 1){
                dfs(grid, row, n-1, m, n);
            }
        }
        for (int col = 0; col < n; col++){
            if (grid[0][col] == 1){
                dfs(grid, 0, col, m, n);
            }
            if (grid[m-1][col] == 1){
                dfs(grid, m-1, col, m, n);
            }
        }
        // count the answer
        for (int row = 0; row < m; row++){
            for (int col = 0; col < n; col++){
                if (grid[row][col] == 1){
                    ans++;
                }
            }
        }
        return ans;
    }
};